### ISSUE TITLE
Deprecate upickle-based pickler

### ISSUE THREAD

**Issue by adamw** [softwaremill/tapir#3401]
It was a fun experiment, but there's too many features which we can't support

**Comment by arturaz** [comment]
Are you talking about https://tapir.softwaremill.com/en/latest/endpoint/pickler.html ? Is there an intended replacement?

**Comment by adamw** [comment]
@arturaz Yes, we're exploring changing the underlying implementation of the json library from uPickle to jsoniter. The end-user interface should remain the same (or very similar)

**Comment by lomigmegard** [comment]
Thanks for the update! 

> So now I'm leaning towards jsoniter, but for that we'd need to change the derivation model (from recursive to "single-shot")

I believe that the full auto-derivation is not something people rely on anymore, I wouldn't see it missing as a problem.

> probably add some AST support

Why does you want an AST for the json part? From outside, the API of Pickler is `Pickler.derived[T]`, with the configuration in scope. The point is to not have to define the Schema and Json codec manually, to keep them in sync. The only place where the underlying library would (should) leak is when defining scalar values. Even there, a dedicated API could be used to bind the Schema with the scalar encoding. 

> Scala-JSON that is fast, secure & integrates well, but unfortunately it doesn't exis

I agree with your comments on the "Scala JSON problem", but at the same time I feel we should be able to hide the library so much that it doesn't matter. To escape these types of problems, I am always tempted to suggest to use a Java library directly, for example jackson. I didn't look at how you implemented Pickler though, the effort to glue the derivation macro might be greater than with a Scala lib.

**Comment by adamw** [comment]
> I believe that the full auto-derivation is not something people rely on anymore, I wouldn't see it missing as a problem.

Well that differs from case to case. But we could start with a single derivation mode (semi-auto) so that `... derives` works as well as top-level givens. Now here various json libraries work differently. If you have a nested AST structure, in some cases you need to derive the codec for all AST members. In some (e.g. jsoniter), you only need to add the derives clause at the top-level AST node. Codecs for children will be then derived as part of this single derivation.

> Why does you want an AST for the json part?

Not that often but often enough, people end up needing fields of type `Json`, which accepts arbitrary (but valid) JSON values. So whatever JSON library we use, it would be good to have it support this kind of functionality.

> for example jackson.

Jackson I think is reflection-based? Something that is very not in the spirit of Scala. Hiding the JSON library would be ideal, yes, although it would also be good to have an option to provide the library-specific configuration, that's not covered by "our" Tapir config. 

But either way: we need to choose a library to hide ;)
